### ISSUE TITLE
[Known Issue] iOSResolver.dll failed to load with Unity 2021.1.11+ on Mac

### ISSUE THREAD

**Issue by chkuang-g** [googlesamples/unity-jar-resolver#441]
### [REQUIRED] Please fill in the following fields:

  * Unity editor version: Issue occurred with `2021.1.11`, `2021.1.12` (Works with `2021.1.10` and `2021.2.0`)
  * External Dependency Manager version: Tried `1.2.160` and `1.2.165`. But version is probably irrelevant here.
  * Source you installed EDM4U: .unitypackage
  * Features in External Dependency Manager in use: iOS Resolver
  * Plugins SDK in use: None
  * Platform you are using the Unity editor on: Mac

### [REQUIRED] Please describe the issue here:
(Please list the full steps to reproduce the issue. Include device logs, Unity logs, and stack traces if available.)

#### Please answer the following, if applicable:
What's the issue repro rate? 100%, even with empty project.

## Issue Statement
iOS Resolver library cannot be loaded in Unity 2021.1.11 and 2021.1.12. (Did not test 2021.2.0 beta yet). It can be loaded properly with Unity 2021.1.10.

After importing EDM4U, Unity logs an error like:
```
Assembly 'Assets/ExternalDependencyManager/Editor/Google.IOSResolver_v1.2.165.dll' will not be loaded due to errors:
Assembly name 'Google.IOSResolver' does not match file name 'Google.IOSResolver_v1.2.165'
```
And iOS Resolver is not loaded, verified through the absence of `Assets > External Dependency Manager > iOS Resolver` menu item. The odd thing is that all the other libraries, ex. Android Resolver, Version Handler, are loaded with no issue.

The same issue has been reported on Unity forum [here](https://forum.unity.com/threads/google-iosresolver-will-not-be-loaded-in-2021-1-11f1-and-later.1128617/)

This seems to be a regression introduced in Unity `2021.1.11` as documented in [the release note](https://unity3d.com/unity/whats-new/2021.1.11).
- Scripting: Editor logs an Error whenever the filename of an assembly does not match its name. ([1109539](https://issuetracker.unity3d.com/issues/scriptable-object-scripts-can-not-be-found-when-the-editor-is-restarted))

## Workaround
Since the issue is the mismatched library name, a simple workaround is to rename the `dll` file.

For instance, 
- Rename `Assets/ExternalDependencyManager/Editor/Google.IOSResolver_v1.2.165.dll` to `Assets/ExternalDependencyManager/Editor/Google.IOSResolver.dll`
- Rename `Assets/ExternalDependencyManager/Editor/Google.IOSResolver_v1.2.165.dll.meta` to `Assets/ExternalDependencyManager/Editor/Google.IOSResolver.dll.meta`
- Rename `Assets/ExternalDependencyManager/Editor/Google.IOSResolver_v1.2.165.dll.mdb` to `Assets/ExternalDependencyManager/Editor/Google.IOSResolver.dll.mdb`
- Rename `Assets/ExternalDependencyManager/Editor/Google.IOSResolver_v1.2.165.dll.mdb.meta` to `Assets/ExternalDependencyManager/Editor/Google.IOSResolver.dll.mdb.meta`

After the renaming, Unity should be able to load iOS Resolver library properly.

**However, there is a caveat**: After renaming, Version Handler will no longer able to manage those files. That is, when a newer version of EDM4U is installed, Version Handler will not be able to automatically remove the older version of iOS Resolver. 

The workaround is to remove the entire folder of `Assets/ExternalDependencyManager/` folder before upgrading.

**Comment by chkuang-g** [comment]
Since this is a Unity issue, the only possible workaround we can provide is to export iOS Resolver library as `Assets/ExternalDependencyManager/Editor/Google.IOSResolver.dll`.  However, as mentioned in the original posts, Version Handler will not be able to remove obsolete libraries automatically.  However, this will work if EDM4U is installed through `.tgz` format.

May need to explore other potential workaround.

**Comment by chkuang-g** [comment]
**Update**:  The issue does NOT occur with Unity `2021.2.0b1`

**Comment by patm1987** [comment]
I am seeing this on Windows with 2021.2.0b1 (installed via Unity Hub) with EDM4U (v1.2.162 installed via the [play games plugin](https://github.com/playgameservices/play-games-plugin-for-unity)).

The suggested workaround does work for me.

I am using Windows 10 (21H1) and my project is on an exFAT file system rather than NTFS.

**Comment by Goncharuk-Nikita** [comment]
+1 also has this issue.
but 2021.2.0b1 does not work for me with Firebase and Facebook SDK.
I downgrade to 2021.1.10.

**Comment by Barokoli** [comment]
+1 also had this issue on 2021.2.0b1
Mac Book M1
Workaround fixed it

**Comment by joaquingrech** [comment]
I´ve just downloaded 166 hoping this was fixed... still same issue
Assembly 'Assets/ExternalDependencyManager/Editor/Google.IOSResolver_v1.2.166.dll' will not be loaded due to errors:
Assembly name 'Google.IOSResolver' does not match file name 'Google.IOSResolver_v1.2.166'

**Comment by gamefish** [comment]
+1 also has this issue.
waiting for the fix. 
do not want to try workaround yet because we are develop under android, will be a few month later to release on iOS

**Comment by chkuang-g** [comment]
Public Issue Tracker from Unity.
https://issuetracker.unity3d.com/issues/only-some-assemblies-fail-to-be-loaded-when-assembly-name-does-not-match-the-file-name

**Comment by itsarjunsinh** [comment]
@chkuang-g Could adding a version code variable inside the dll be a solution? It's a shot in the dark, but I suppose you can then adhere to new naming conventions and still allow the version handler to manage files.
A fix for this would be great. Renaming files is particularly tedious for me since I use tarballs for dependencies. I've added a duplicate EDM via Unity package for now.

**Comment by LaCreArthur** [comment]
Renaming the file manually did the trick, thanks, but it should not be that way

**Comment by Thaina** [comment]
Are there any progress on this? What is the actual solution that will came out in future release? Is it the problem of unity itself or this google repo?

It seem this problem would affect firebase and facebook (and any other library depend on EDM4U) as well. And so we can't update to 2021.2b

**Comment by chkuang-g** [comment]
Hey folks,

**TL;DR;**  It is unfortunately Unity did not take any action to fix this from their side over 3 months. :( The quickest workaround is still renaming `IOSResolver_v1.2.165.dll` to `IOSResolver.dll`.  This is a pretty tricky to create a workaround in EDM4U which works for both `.unitypackage` and `.tgz` format.  I'm going to explain why and propose a potential workaround in EDM4U.

## Why versions in filename?

The reason why we are adding a postfix like `_v1.2.165` in the file name is to make sure when you **upgrade** EDM4U through **`.unitypackage`**, directly or indirectly, Version Handler in EDM4U is able to enable the later version and remove the older version. ([brief description here](https://github.com/googlesamples/unity-jar-resolver#unity-plugin-version-management)). Without this postfix, whatever imported later will override the existing files in your project. For instance, if you import packages in the following order:
1. Firebase `8.0.0` (EDM4U `1.2.166`)
2. Facebook `11.0.0` (EDM4U `1.2.135`)

Without the postfix, you will end-up having EDM4U `1.2.135` in your project since the same file name imported later override the existing one in the project. With the postfix, after you import Facebook SDK, EDM4U can kick-in and remove EDM4U `1.2.135` without you doing anything.

Note that versioning is not an issue if you use `.tgz` through [UPM](https://docs.unity3d.com/Packages/com.unity.package-manager-ui@1.8/manual/index.html), since Unity basically removes the older version before they install the newer one.

## Potential Workaround in EDM4U

Version information is not only stored in the filename but also in the [Asset label](https://github.com/googlesamples/unity-jar-resolver/blob/41072ec47290642e126388ae3c5dd43d42695d9b/exploded/Assets/ExternalDependencyManager/Editor/Google.IOSResolver_v1.2.166.dll.meta#L4) and in [code](https://github.com/googlesamples/unity-jar-resolver/blob/30b4f9292db3bd11340883dd80872c9852ae70e3/source/IOSResolver/src/VersionNumber.cs#L30)

Since `.unitypackage` and `.tgz` are two different stories, I will propose two different workaround in EDM4U.

- `.tgz`: This should be "simpler". Basically we can ship all EDM4U libraries without version postfix since they are not used at all. This should be some change in our [build](https://github.com/googlesamples/unity-jar-resolver/blob/master/build.gradle#L824) and release process.
- `.unitypackage`: As explained above, we HAVE to keep the version postfix for the `dll` shipped in `.unitypackage`, or it would be equally painful experience for you guys when you upgrade. One possible workaround we can add to Version Handler is: 
  - The `.dll` in `.unitypackage` will still have version in the filename
  - After a new versioned file is imported, EDM4U can compare all same files (different versions) in the Unity project, find the latest version of the file, and rename that file to remove the version part. That is, if both `IOSResolver.dll` (1.2.100) and `IOSResolver_v1.2.500.dll` exists in the project, change `IOSResolver.dll` to `IOSResolver_v1.2.100.dll` and disable it, and rename `IOSResolver_v1.2.500.dll` into `IOSResolver.dll`.

Will keep you updated.

Shawn

**Comment by Thaina** [comment]
Thank you very much for your clarification

One workaround I could think of is, instead of having version as postfix in dll, you might have version as postfix in the folder, and remove the whole folder for each version. With that we then don't mess with the dll filename and it would also not a problem for tgz import isn't it?

**Comment by Thaina** [comment]
Also there is a weird error in 2021.2 and 2022.1, I have made this post in forum

https://forum.unity.com/threads/error-on-import-google-external-dependency-manager-and-firebase-package.1167353/

Is it related?

**Comment by chkuang-g** [comment]
@Thaina Thank you for the suggestion! Use version as folder name is an interesting thought.  Let me explore a bit to this route.

As for you post, at least we know `IOSResolver` cannot be loaded properly in 2021.1.11+ and 2022. And Firebase editor tool uses EDM4U APIs.  It is logical that `Firebase.Editor.dll` cannot be loaded. :/

You can try to rename See if renaming those libraries, restart the editor and see if the error goes away.  If Firebase editor tool can be loaded, you should be able to find `Windows > Firebase > Documentation` menu item.

**Comment by chkuang-g** [comment]
Alright, I created a couple of internal tasks to track the progress for this issue.

- b/191718824 - Master task
- b/199452980 - Unblock `.tgz`
- b/199453559 - Unblock `.unitypackage` or both.

**Comment by Thaina** [comment]
Have rename dll manually and repackage it into tgz cause these errors in firebase

![image](https://user-images.githubusercontent.com/1042507/133053749-042a7df0-c939-49fe-98a9-8ce5e058e543.png)

**Comment by chkuang-g** [comment]
I did some experiment with Unity `2021.2.1b11.3437`. What I found is that this version of Unity basically does not load ANY `.dll` from Unity `.tgz` package.  :(
I tried the pre-built EDM4U from [here](https://developers.google.com/unity/archive#external_dependency_manager_for_unity) and another version built from scratch.  I have no idea what is wrong with this beta version Unity yet. Might need to dig their release note and see if there is any noticeable breaking changes.

On the other hand, Unity `2021.1.21f1` is more predictable.  I was able to rename ALL `.dll` in `.tgz` package and all libraries were loaded properly.

At this point, I would like to skip `2021.2.1b` since it is a beta version, and just focus on unblocking those who are using non-beta Unity version.

**Comment by Thaina** [comment]
I see. Thank you very much

Also the 2022.1a (current alpha) face the same issue as 2021.2b. That was something wrong for sure

**Comment by Linnun** [comment]
I don't see this error anymore with the update to 2021.2.1b12. Might be fixed?

**Comment by ohhh-source-code** [comment]
renaming to IOSResolver.dll for all four files only created another error, which is 

Assembly 'Assets/ExternalDependencyManager/Editor/Google.IOSResolver.dll' will not be loaded due to errors:
Unable to resolve reference 'UnityEditor.iOS.Extensions.Xcode'. Is the assembly missing or incompatible with the current platform?
Reference validation can be disabled in the Plugin Inspector.

Any idea?

**Comment by chkuang-g** [comment]
@ohhh-source-code `UnityEditor.iOS.Extensions.Xcode` is a Unity API used by `IOSResolver`.  If it can not be resolved, is it possible that you have not install iOS build support?

**Comment by chkuang-g** [comment]
Also, `1.2.167` was released

**Comment by Volcore** [comment]
@chkuang-g I've just tried this, and it's not working. I think the problem is that the libraries are now just completely missing from the .tgz version. The unitypackage version seems to still have them:

<img width="472" alt="image" src="https://user-images.githubusercontent.com/95790/136672519-66b36beb-1058-482e-8346-10fb06c9d140.png">
